[CMake/stacktrace] Undefined reference to symbol

Recent precice develop, after merging cmake overhaul:

Configuring with `PETSC=off`
```
 % cmake -DPETSC=off ../.. 
-- The CXX compiler identification is GNU 8.2.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build configuration: Debug
-- Special platform settings: none
-- --------------------
-- == CONFIGURATION
-- Library version to build: 1.3.0
-- Build configuration: Debug
-- Build shared libraries: <undefined>
-- Target system: Linux-4.20.7-arch1-1-ARCH
-- Host system: Linux-4.20.7-arch1-1-ARCH
-- CXX compiler: /usr/bin/c++
-- CXX compiler flags: <empty>
-- CXX linker: /usr/bin/ld
-- Install prefix: /usr/local
-- Source directory: /home/lindnefn/precice
-- Binary directory: /home/lindnefn/precice/build/last
-- --------------------
-- Build with MPI: ON
-- Build with PETSC: off
-- Build with PYTHON: ON
-- Special platform settings: none
-- Configure package generation: ON
-- Install tests/testfiles: ON
-- --------------------
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   filesystem
--   log
--   log_setup
--   program_options
--   system
--   thread
--   unit_test_framework
--   date_time
--   regex
--   chrono
--   atomic
-- Found Eigen3: /usr/include/eigen3 (Required is at least version "3.2") 
-- Found LibXml2: /usr/lib/libxml2.so (found version "2.9.9") 
-- Found MPI_CXX: /usr/lib/openmpi/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- PETSc support disabled
-- Found PythonLibs: /usr/lib/libpython2.7.so (found suitable version "2.7.15", minimum required is "2.7") 
-- Found PythonInterp: /usr/bin/python2.7 (found version "2.7.15") 
-- Found PythonLibs: /usr/lib/libpython2.7.so (found version "2.7.15") 
-- Found NumPy: /usr/lib/python2.7/site-packages/numpy/core/include (found version "1.16.1") 
-- Compressing changelog
-- Compressing manpage: /home/lindnefn/precice/build/last/packaging/man1/binprecice.1
-- Compressing manpage: /home/lindnefn/precice/build/last/packaging/man1/testprecice.1
-- Configuring done
-- Generating done
-- Build files have been written to: /home/lindnefn/precice/build/last
```

```
% make -j2
[...]
[ 99%] Building CXX object CMakeFiles/testprecice.dir/src/mapping/tests/NearestProjectionMappingTest.cpp.o
[100%] Linking CXX executable testprecice
/usr/bin/ld: libprecice.a(stacktrace.cpp.o): undefined reference to symbol 'dladdr@@GLIBC_2.2.5'
/usr/bin/ld: /usr/lib/libdl.so.2: error adding symbols: DSO missing from command line
collect2: Fehler: ld gab 1 als Ende-Status zurück
make[2]: *** [CMakeFiles/testprecice.dir/build.make:1075: testprecice] Fehler 1
make[1]: *** [CMakeFiles/Makefile2:147: CMakeFiles/testprecice.dir/all] Fehler 2
make: *** [Makefile:163: all] Fehler 2
```
Configuring with `PETSC=on`, i.e. without options:
```
% cmake ../..
-- The CXX compiler identification is GNU 8.2.1
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build configuration: Debug
-- Special platform settings: none
-- --------------------
-- == CONFIGURATION
-- Library version to build: 1.3.0
-- Build configuration: Debug
-- Build shared libraries: <undefined>
-- Target system: Linux-4.20.7-arch1-1-ARCH
-- Host system: Linux-4.20.7-arch1-1-ARCH
-- CXX compiler: /usr/bin/c++
-- CXX compiler flags: <empty>
-- CXX linker: /usr/bin/ld
-- Install prefix: /usr/local
-- Source directory: /home/lindnefn/precice
-- Binary directory: /home/lindnefn/precice/build/cmake-petsc
-- --------------------
-- Build with MPI: ON
-- Build with PETSC: ON
-- Build with PYTHON: ON
-- Special platform settings: none
-- Configure package generation: ON
-- Install tests/testfiles: ON
-- --------------------
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.68.0
-- Found the following Boost libraries:
--   filesystem
--   log
--   log_setup
--   program_options
--   system
--   thread
--   unit_test_framework
--   date_time
--   regex
--   chrono
--   atomic
-- Found Eigen3: /usr/include/eigen3 (Required is at least version "3.2") 
-- Found LibXml2: /usr/lib/libxml2.so (found version "2.9.9") 
-- Found MPI_CXX: /usr/lib/openmpi/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- petsc_lib_dir /home/lindnefn/software/petsc/arch-linux2-c-debug/lib
-- Recognized PETSc install with single library for all packages
-- Performing Test MULTIPASS_TEST_1_petsc_works_minimal
-- Performing Test MULTIPASS_TEST_1_petsc_works_minimal - Failed
-- Performing Test MULTIPASS_TEST_2_petsc_works_allincludes
-- Performing Test MULTIPASS_TEST_2_petsc_works_allincludes - Failed
-- Performing Test MULTIPASS_TEST_3_petsc_works_alllibraries
-- Performing Test MULTIPASS_TEST_3_petsc_works_alllibraries - Success
-- PETSc only need minimal includes, but requires explicit linking to all dependencies.  This is expected when PETSc is built with static libraries.
-- Found PETSc: /home/lindnefn/software/petsc/arch-linux2-c-debug/include;/home/lindnefn/software/petsc/include (found suitable version "3.10.3", minimum required is "3.6") 
-- Found PythonLibs: /usr/lib/libpython2.7.so (found suitable version "2.7.15", minimum required is "2.7") 
-- Found PythonInterp: /usr/bin/python2.7 (found version "2.7.15") 
-- Found PythonLibs: /usr/lib/libpython2.7.so (found version "2.7.15") 
-- Found NumPy: /usr/lib/python2.7/site-packages/numpy/core/include (found version "1.16.1") 
-- Compressing changelog
-- Compressing manpage: /home/lindnefn/precice/build/cmake-petsc/packaging/man1/binprecice.1
-- Compressing manpage: /home/lindnefn/precice/build/cmake-petsc/packaging/man1/testprecice.1
-- Configuring done
-- Generating done
-- Build files have been written to: /home/lindnefn/precice/build/cmake-petsc
```
compiles fine.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -84,6 +84,11 @@ find_package (Threads REQUIRED)
 find_package(Boost 1.65.1 REQUIRED
   COMPONENTS filesystem log log_setup program_options system thread unit_test_framework)
 
+if (UNIX OR APPLE OR MINGW)
+  # Used for boost stacktrace on POSIX and MINGW
+  find_package(DL REQUIRED)
+endif()
+
 find_package(Eigen3 3.2 REQUIRED)
 
 find_package(LibXml2 REQUIRED)
@@ -145,6 +150,9 @@ set_target_properties(precice PROPERTIES
 target_compile_definitions(precice PRIVATE BOOST_ALL_DYN_LINK BOOST_ASIO_ENABLE_OLD_SERVICES)
 target_link_libraries(precice PRIVATE ${Boost_LIBRARIES})
 target_include_directories(precice PRIVATE ${Boost_INCLUDE_DIRS})
+if(UNIX OR APPLE OR MINGW)
+  target_link_libraries(precice PRIVATE DL::DL)
+endif()
 
 # Setup Eigen3
 target_link_libraries(precice PRIVATE Eigen3::Eigen)

--- a/cmake/modules/FindDL.cmake
+++ b/cmake/modules/FindDL.cmake
@@ -1,0 +1,34 @@
+# This module defines the following variables:
+#
+# DL_FOUND
+#   True if libdl found.
+#
+# DL_LIBRARIES
+#   The library libdl
+#
+# This module exports the following IMPORTED target:
+#
+# DL::DL
+#
+
+find_library(DL_LIBRARY
+    NAMES dl libdl
+    HINTS ${DL_ROOT} $ENV{DL_ROOT})
+
+include(FindPackageHandleStandardArgs)
+find_package_handle_standard_args(DL
+    REQUIRED_VARS DL_LIBRARY)
+
+if(DL_FOUND)
+    set(DL_LIBRARIES ${DL_LIBRARY})
+
+    if(NOT TARGET DL::DL)
+        add_library(DL::DL UNKNOWN IMPORTED)
+        set_target_properties(DL::DL PROPERTIES
+            INTERFACE_LINK_LIBRARIES "${DL_LIBRARY}"
+            )
+        set_property(TARGET DL::DL APPEND PROPERTY IMPORTED_LOCATION "${DL_LIBRARY}")
+    endif()
+endif()
+
+mark_as_advanced(DL_LIBRARIES)

